---
title: "Practice Activity 5.1: JSON Data Format and APIs"
format:
  html:
    embed-resources: true
---

# Hierarchical Data, the JSON Data Format, and APIs

In [1]:
import pandas as pd

## Shows Data

First we'll work with the "Girls" shows JSON data from the reading.

In [2]:
# Fetch data from a URL
import requests
response = requests.get("https://dlsun.github.io/pods/data/tvshows.json")

import json
data_shows = response.json()

In [3]:
df_shows = pd.json_normalize(data_shows)
df_shows

,id,url,name,type,language,genres,status,runtime,premiered,officialSite,...,externals.thetvdb,externals.imdb,image.medium,image.original,network,webChannel.id,webChannel.name,webChannel.country.name,webChannel.country.code,webChannel.country.timezone
0,139,http://www.tvmaze.com/shows/139/girls,Girls,Scripted,English,"[Drama, Romance]",Ended,30,2012-04-15,http://www.hbo.com/girls,...,220411,tt1723816,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
1,722,http://www.tvmaze.com/shows/722/the-golden-girls,The Golden Girls,Scripted,English,"[Drama, Comedy]",Ended,30,1985-09-14,None,...,71292,tt0088526,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
2,23542,http://www.tvmaze.com/shows/23542/good-girls,Good Girls,Scripted,English,"[Drama, Comedy, Crime]",Running,60,2018-02-26,https://www.nbc.com/good-girls?nbc=1,...,328577,tt6474378,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
3,6771,http://www.tvmaze.com/shows/6771/the-powerpuff...,The Powerpuff Girls,Animation,English,"[Comedy, Action, Science-Fiction]",Running,15,2016-04-04,https://www.cartoonnetwork.com/video/powerpuff...,...,307473,tt4718304,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
4,42726,http://www.tvmaze.com/shows/42726/florida-girls,Florida Girls,Scripted,English,[Comedy],Running,30,2019-07-10,https://poptv.com/floridagirls,...,363682,tt8548870,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
5,32087,http://www.tvmaze.com/shows/32087/chicken-girls,Chicken Girls,Scripted,English,"[Drama, Children, Music]",Running,16,2017-09-05,https://www.youtube.com/playlist?list=PLVewHiZ...,...,339854,None,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,274.0,Brat,United States,US,America/New_York
6,33320,http://www.tvmaze.com/shows/33320/derry-girls,Derry Girls,Scripted,English,[Comedy],Running,30,2018-01-04,http://www.channel4.com/programmes/derry-girls,...,338903,tt7120662,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
7,1955,http://www.tvmaze.com/shows/1955/the-powerpuff...,The Powerpuff Girls,Animation,English,"[Action, Children, Crime]",Ended,30,1998-11-18,None,...,76200,tt0175058,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
8,1073,http://www.tvmaze.com/shows/1073/bomb-girls,Bomb Girls,Scripted,English,"[Drama, Romance, War]",Ended,60,2012-01-04,None,...,254378,tt1955311,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN
9,525,http://www.tvmaze.com/shows/525/gilmore-girls,Gilmore Girls,Scripted,English,"[Drama, Comedy, Romance]",Ended,60,2000-10-05,None,...,76568,tt0238784,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,NaN,NaN,NaN,NaN


1\. Summarize the networks represented by these shows and the number of these shows that aired on each network.

In [4]:
df_shows["network.name"].value_counts()

network.name
NBC                2
Cartoon Network    2
HBO                1
Pop                1
Channel 4          1
Global             1
The CW             1
Name: count, dtype: int64

2\. Find the number of seasons for each show in the data. Do this two ways: one which uses `df_shows`, and another that first flattens `data_shows` to a different data frame.

In [5]:
df_shows["num_seasons"] = df_shows["seasons"].apply(len)

df_shows[["name", "num_seasons"]]

,name,num_seasons
0,Girls,6
1,The Golden Girls,7
2,Good Girls,3
3,The Powerpuff Girls,3
4,Florida Girls,1
5,Chicken Girls,5
6,Derry Girls,2
7,The Powerpuff Girls,6
8,Bomb Girls,2
9,Gilmore Girls,8


In [6]:
df_seasons = pd.json_normalize(
    data_shows,
    record_path="seasons",
    meta=["name"],
    meta_prefix="show_"
)

df_seasons.groupby(["id", "show_name"]).size().reset_index

<bound method Series.reset_index of id      show_name          
650     Girls                  1
651     Girls                  1
652     Girls                  1
653     Girls                  1
2080    Gilmore Girls          1
2081    Gilmore Girls          1
2082    Gilmore Girls          1
2083    Gilmore Girls          1
2084    Gilmore Girls          1
2085    Gilmore Girls          1
2086    Gilmore Girls          1
2923    The Golden Girls       1
2924    The Golden Girls       1
2925    The Golden Girls       1
2926    The Golden Girls       1
2927    The Golden Girls       1
2928    The Golden Girls       1
2929    The Golden Girls       1
4998    Bomb Girls             1
4999    Bomb Girls             1
7073    The Powerpuff Girls    1
7074    The Powerpuff Girls    1
7075    The Powerpuff Girls    1
7076    The Powerpuff Girls    1
7077    The Powerpuff Girls    1
7078    The Powerpuff Girls    1
20323   The Powerpuff Girls    1
21282   Girls                  1
34116   Gilm

3\. For each episode, find the length (number of characters) of the title. Then create summaries to answer: Which show tends to have the longest episode titles? The shortest?

In [7]:
df_episodes = pd.json_normalize(
    data_shows,
    record_path=["seasons", "episodes"],
    meta=["name"],
    meta_prefix="show_")

df_episodes

,id,url,name,season,number,airdate,airtime,airstamp,runtime,summary,image.medium,image.original,image,show_name
0,10820,http://www.tvmaze.com/episodes/10820/girls-1x0...,Pilot,1,1.0,2012-04-15,22:30,2012-04-16T02:30:00+00:00,30,<p>In the premiere of this comedy about twenty...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,Girls
1,10821,http://www.tvmaze.com/episodes/10821/girls-1x0...,Vagina Panic,1,2.0,2012-04-22,22:30,2012-04-23T02:30:00+00:00,30,<p>An appointment at a women's clinic doesn't ...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,Girls
2,10822,http://www.tvmaze.com/episodes/10822/girls-1x0...,All Adventurous Women Do,1,3.0,2012-04-29,22:30,2012-04-30T02:30:00+00:00,30,<p>Hannah contacts her college boyfriend to fi...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,Girls
3,10823,http://www.tvmaze.com/episodes/10823/girls-1x0...,Hannah's Diary,1,4.0,2012-05-06,22:30,2012-05-07T02:30:00+00:00,30,<p>Adam's risqué text message sends Hannah ove...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,Girls
4,10824,http://www.tvmaze.com/episodes/10824/girls-1x0...,Hard Being Easy,1,5.0,2012-05-13,22:30,2012-05-14T02:30:00+00:00,30,<p>Hannah tries a different tack with her boss...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,Girls
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
736,47635,http://www.tvmaze.com/episodes/47635/gilmore-g...,Hay Bale Maze,7,18.0,2007-04-17,21:00,2007-04-18T01:00:00+00:00,60,<p>Brief encounter: Stars Hollow's centerpiece...,NaN,NaN,NaN,Gilmore Girls
737,47636,http://www.tvmaze.com/episodes/47636/gilmore-g...,It's Just Like Riding a Bike,7,19.0,2007-04-24,21:00,2007-04-25T01:00:00+00:00,60,<p>Lorelai needs help shopping for a new car. ...,NaN,NaN,NaN,Gilmore Girls
738,47637,http://www.tvmaze.com/episodes/47637/gilmore-g...,Lorelai? Lorelai?,7,20.0,2007-05-01,21:00,2007-05-02T01:00:00+00:00,60,"<p>On karaoke night, a tipsy Lorelai take the ...",NaN,NaN,NaN,Gilmore Girls
739,47638,http://www.tvmaze.com/episodes/47638/gilmore-g...,Unto the Breach,7,21.0,2007-05-08,21:00,2007-05-09T01:00:00+00:00,60,<p>Pomp and circumstances. Rory's graduation b...,NaN,NaN,NaN,Gilmore Girls


In [8]:
df_episodes.columns

Index(['id', 'url', 'name', 'season', 'number', 'airdate', 'airtime',
       'airstamp', 'runtime', 'summary', 'image.medium', 'image.original',
       'image', 'show_name'],
      dtype='object')

In [9]:
df_episodes["title_length"] = df_episodes["name"].str.len()

In [10]:
df_episodes.groupby("show_name")["title_length"].mean().sort_values()

show_name
Derry Girls            12.500000
Girls                  13.111111
Bomb Girls             14.421053
Good Girls             15.461538
Chicken Girls          16.434211
The Golden Girls       19.723757
Florida Girls          20.400000
The Powerpuff Girls    21.184080
Gilmore Girls          21.431373
Name: title_length, dtype: float64

4\. Do any cast members in the data set share a birthday with you? Who, and what show are they on? (If there isn't anyone, try a different day.)

In [11]:
df_cast = pd.json_normalize(
    data_shows,
    record_path="cast",
    meta=["name"],
    meta_prefix="show_")

df_cast

,self,voice,person.id,person.url,person.name,person.country.name,person.country.code,person.country.timezone,person.birthday,person.deathday,...,person.image.medium,person.image.original,character.id,character.url,character.name,character.image.medium,character.image.original,person.country,character.image,show_name
0,False,False,27410,http://www.tvmaze.com/people/27410/lena-dunham,Lena Dunham,United States,US,America/New_York,1986-05-13,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,36886,http://www.tvmaze.com/characters/36886/girls-h...,Hannah Horvath,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Girls
1,False,False,11102,http://www.tvmaze.com/people/11102/allison-wil...,Allison Williams,United States,US,America/New_York,1988-04-13,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,36887,http://www.tvmaze.com/characters/36887/girls-m...,Marnie Michaels,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Girls
2,False,False,27411,http://www.tvmaze.com/people/27411/jemima-kirke,Jemima Kirke,United Kingdom,GB,Europe/London,1985-04-26,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,36888,http://www.tvmaze.com/characters/36888/girls-j...,Jessa Johansson,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Girls
3,False,False,16405,http://www.tvmaze.com/people/16405/zosia-mamet,Zosia Mamet,United States,US,America/New_York,1988-02-02,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,36889,http://www.tvmaze.com/characters/36889/girls-s...,Shoshanna Shapiro,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Girls
4,False,False,24858,http://www.tvmaze.com/people/24858/adam-driver,Adam Driver,United States,US,America/New_York,1983-11-19,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,36890,http://www.tvmaze.com/characters/36890/girls-a...,Adam Sackler,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Girls
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
107,False,False,20369,http://www.tvmaze.com/people/20369/edward-herr...,Edward Herrmann,United States,US,America/New_York,1943-07-21,2014-12-31,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,85545,http://www.tvmaze.com/characters/85545/gilmore...,Richard Gilmore,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Gilmore Girls
108,False,False,3354,http://www.tvmaze.com/people/3354/jared-padalecki,Jared Padalecki,United States,US,America/New_York,1982-07-19,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,85552,http://www.tvmaze.com/characters/85552/gilmore...,Dean Forester,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Gilmore Girls
109,False,False,18121,http://www.tvmaze.com/people/18121/matt-czuchry,Matt Czuchry,United States,US,America/New_York,1977-05-20,None,...,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,85550,http://www.tvmaze.com/characters/85550/gilmore...,Logan Huntzberger,http://static.tvmaze.com/uploads/images/medium...,http://static.tvmaze.com/uploads/images/origin...,NaN,NaN,Gilmore Girls
110,False,False,20026,http://www.tvmaze.com/people/20026/milo-ventim...,Milo Ventimiglia,United States,US,America/New_York,1977-07-08,None,...,http://static.tvmaze.com/uploads/images/me

In [12]:
df_cast[
    df_cast["person.birthday"].str.endswith("-06-30", na=False)
][["person.name", "person.birthday", "show_name"]]

,person.name,person.birthday,show_name
83,Peter Outerbridge,1966-06-30,Bomb Girls


## TVMaze API

Now you will work with the [TVMaze API](http://www.tvmaze.com/api) from the reading. Use the API to request JSON data that you can use to answer the following questions.

1\. What was the longest show that aired in the U.S. on February 4, 2018?

_Hint:_ Use the ["Schedule" endpoint](http://www.tvmaze.com/api#schedule) to first get the data for all shows that aired on that date.

In [13]:
import requests

response = requests.get(
    "https://api.tvmaze.com/schedule",
    params={
        "country": "US",
        "date": "2018-02-04"
    }
)

data_schedule = response.json()

In [14]:
df_schedule = pd.json_normalize(data_schedule)

df_schedule.head()

,id,url,name,season,number,type,airdate,airtime,airstamp,runtime,...,image.medium,image.original,show.network,show.webChannel.id,show.webChannel.name,show.webChannel.country.name,show.webChannel.country.code,show.webChannel.country.timezone,show.webChannel.officialSite,show.image
0,1442787,https://www.tvmaze.com/episodes/1442787/1st-lo...,Going for Gold in PyeongChang,9,14.0,regular,2018-02-03,01:02,2018-02-04T06:02:00+00:00,30,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1388511,https://www.tvmaze.com/episodes/1388511/americ...,Ultimate Italian,18,5.0,regular,2018-02-03,02:30,2018-02-04T07:30:00+00:00,30,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,1387556,https://www.tvmaze.com/episodes/1387556/marvel...,"Spider-Island, Part One",1,19.0,regular,2018-02-04,07:00,2018-02-04T12:00:00+00:00,30,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,1387558,https://www.tvmaze.com/episodes/1387558/marvel...,"Spider-Island, Part Two",1,20.0,regular,2018-02-04,07:30,2018-02-04T12:30:00+00:00,30,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,1390384,https://www.tvmaze.com/episodes/1390384/ben-10...,High Stress Express,2,11.0,regular,2018-02-04,13:00,2018-02-04T13:00:00+00:00,11,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [15]:
df_schedule.sort_values("runtime", ascending=False)[["show.name","runtime","airtime"]].head(2)

,show.name,runtime,airtime
23,Super Bowl,210,18:30
29,Alaska: The Last Frontier,180,20:00


2\. Among all shows that aired in the U.S. on Feburary 4, 2018, which non-voice actors appeared on more than one show? Note: some people are credited with multiple rows on the same show and thus appear as multiple rows in the data frame.

Hint: You will need to write a for loop to make multiple requests. Use the "show.id" from the data from part 1, and use the ["Shows" endpoint](http://www.tvmaze.com/api#show-cast) to get the cast of each show. Don't forget to stagger your requests, or you will be blocked by the website!

In [16]:
show_ids = df_schedule["show.id"].unique()

show_ids

array([ 7795,  5599, 21928, 22084,  8835, 15779,  7249, 72487,  7250,
        6743,  2984, 11436, 17116,  8834, 10636, 15260,  6821, 32095,
       17584,  9659,  6011,  9539,  4022, 70137, 89260,  6456,  1195,
       33430, 28273,  4869,  8999,  5965, 34214, 17128, 12069,  1951,
         686, 23065])

In [17]:
## https://api.tvmaze.com/shows/1/cast
import time


cast_records = []
for show_id in show_ids:
    response = requests.get(f"https://api.tvmaze.com/shows/{show_id}/cast")
    if response.status_code != 200:
        print("Skipped", show_id, "State error", response.status_code)
        continue 
    members = response.json()
    for member in members:
        member["show_id"] = int(show_id)
    cast_records.extend(members)
    time.sleep(0.5)

In [18]:
import time

cast = []

for show_id in show_ids:

    response = requests.get(
        f"https://api.tvmaze.com/shows/{show_id}/cast"
    )

    members = response.json()

    for m in members:
        m["show_id"] = show_id

    cast.extend(members)

    time.sleep(0.5)

In [19]:
df_cast = pd.json_normalize(cast)

df_cast.head()

,self,voice,show_id,person.id,person.url,person.name,person.country.name,person.country.code,person.country.timezone,person.birthday,...,person._links.self.href,character.id,character.url,character.name,character.image.medium,character.image.original,character._links.self.href,person.country,person.image,character.image
0,True,False,7795,117627,https://www.tvmaze.com/people/117627/johnny-de...,Johnny Devenanzio,United States,US,America/New_York,1982-06-22,...,https://api.tvmaze.com/people/117627,627654,https://www.tvmaze.com/characters/627654/1st-l...,Host,https://static.tvmaze.com/uploads/images/mediu...,https://static.tvmaze.com/uploads/images/origi...,https://api.tvmaze.com/characters/627654,NaN,NaN,NaN
1,True,False,7795,132602,https://www.tvmaze.com/people/132602/angela-sun,Angela Sun,NaN,NaN,NaN,None,...,https://api.tvmaze.com/people/132602,627644,https://www.tvmaze.com/characters/627644/1st-l...,Host,https://static.tvmaze.com/uploads/images/mediu...,https://static.tvmaze.com/uploads/images/origi...,https://api.tvmaze.com/characters/627644,NaN,NaN,NaN
2,True,False,7795,71115,https://www.tvmaze.com/people/71115/ashley-rob...,Ashley Roberts,United States,US,America/New_York,1981-09-14,...,https://api.tvmaze.com/people/71115,627646,https://www.tvmaze.com/characters/627646/1st-l...,Host,https://static.tvmaze.com/uploads/images/mediu...,https://static.tvmaze.com/uploads/images/origi...,https://api.tvmaze.com/characters/627646,NaN,NaN,NaN
3,True,False,7795,51988,https://www.tvmaze.com/people/51988/audrina-pa...,Audrina Patridge,United States,US,America/New_York,1985-05-09,...,https://api.tvmaze.com/people/51988,251236,https://www.tvmaze.com/characters/251236/1st-l...,Host,https://static.tvmaze.com/uploads/images/mediu...,https://static.tvmaze.com/uploads/images/origi...,https://api.tvmaze.com/characters/251236,NaN,NaN,NaN
4,True,False,7795,235612,https://www.tvmaze.com/people/235612/pedro-and...,Pedro Andrade,NaN,NaN,NaN,None,...,https://api.tvmaze.com/people/235612,627647,https://www.tvmaze.com/characters/627647/1st-l...,Host,https://static.tvmaze.com/uploads/images/mediu...,https://static.tvmaze.com/uploads/images/origi...,https://api.tvmaze.com/characters/627647,NaN,NaN,NaN


In [20]:
df_nonvoice = df_cast[
    df_cast["voice"] == False]

In [21]:
# Unique actors only
actor_shows = df_nonvoice[
    ["person.name", "show_id"]].drop_duplicates()

In [22]:
actor_counts = (
    actor_shows
    .groupby("person.name")
    .size()
    .sort_values(ascending=False))

actor_counts

person.name
Tom Llamas             2
John Dickerson         2
Adam Ried              1
Major Garrett          1
Mandy Moore            1
                      ..
Fareed Zakaria         1
Gabriele Bertaccini    1
Gardner Loulan         1
Garrett Hilbert        1
Yvonne Ruperti         1
Length: 182, dtype: int64

In [23]:
actor_counts[
    actor_counts > 1]

person.name
Tom Llamas        2
John Dickerson    2
dtype: int64

In [24]:
# after the loop
df_cast = pd.json_normalize(cast)

df_nonvoice = df_cast[
    df_cast["voice"] == False
]

actor_shows = df_nonvoice[
    ["person.name", "show_id"]
].drop_duplicates()

actor_counts = (
    actor_shows
    .groupby("person.name")
    .size()
    .sort_values(ascending=False)
)

actor_counts[
    actor_counts > 1
]

person.name
Tom Llamas        2
John Dickerson    2
dtype: int64

# Tasty API

[Tasty.co](http://tasty.co) is a website and app that offers food recipes. They have made these recipes available through a [a REST API](https://rapidapi.com/apidojo/api/tasty). However, unlike the TVMaze API, this one requires authentication.

Specifically, you will need to create an account and subscribe to the "Basic" (free) plant. You will then be provided with an API key (X-Rapid-API-Key) that will need to be supplied with every request you make. This is used to track and limit usage.

1. [create an account](https://rapidapi.com/apidojo/api/tasty) and subscribe to the "Basic" (free) plan
2. log in and copy the X-RapidAPI-Key, which is a long string of letters and digits
3. paste this key to replace "PUT-YOUR-KEY-HERE" in the `headers` below

If you did everything correctly, then running the cell below should return a JSON object containing all the tags recognized by the Tasty API.

In [30]:
headers = {
    "x-rapidapi-host": "tasty.p.rapidapi.com",
    "x-rapidapi-key": "1c7cf89b7dmshf38ae1b93083922p1d15b4jsn58a4cfc9debf"
}

In [36]:
import requests

url = "https://tasty.p.rapidapi.com/recipes/list"

querystring = {"q":"daikon","tags":"under_30_minutes","size":"20","from":"0"}

headers = {
	"x-rapidapi-key": "1c7cf89b7dmshf38ae1b93083922p1d15b4jsn58a4cfc9debf",
	"x-rapidapi-host": "tasty.p.rapidapi.com"
}

response = requests.get(url, headers=headers, params=querystring)

print(response.json())

{'count': 4, 'results': [{'nutrition_visibility': 'auto', 'country': 'US', 'instructions': [{'start_time': 9666, 'appliance': None, 'end_time': 12500, 'temperature': None, 'id': 40558, 'position': 1, 'display_text': 'Sriracha aioli: Combine both ingredients in a small bowl and set aside.'}, {'start_time': 0, 'appliance': None, 'end_time': 0, 'temperature': None, 'id': 40559, 'position': 2, 'display_text': 'Carrot and daikon pickles: Combine carrots, radish, sugar, and salt in a large bowl. Using fingertips, massage salt and sugar into vegetables until dissolved. Add water and rice vinegar. Set aside.'}, {'start_time': 13666, 'appliance': None, 'end_time': 26450, 'temperature': None, 'id': 40562, 'position': 3, 'display_text': 'Burgers: Combine the pork, garlic, scallions, fish sauce, chili sauce, sugar, pepper, and salt in a large bowl. Mix gently until spices are evenly distributed, but don’t overwork meat. Form into four balls.'}, {'start_time': 26950, 'appliance': None, 'hacks': [{'

In [37]:
# Convert the API response to JSON
data_daikon = response.json()

df_daikon = pd.json_normalize(
    data_daikon["results"])

df_daikon

df_daikon.columns

Index(['nutrition_visibility', 'country', 'instructions', 'keywords', 'facebook_posts', 'language', 'seo_path', 'id',
       'brand', 'slug', 'servings_noun_singular', 'show_id', 'prep_time_minutes', 'is_subscriber_content', 'sections',
       'brand_id', 'tags', 'is_app_only', 'name', 'compilations', 'num_servings', 'buzz_id',
       'tips_and_ratings_enabled', 'aspect_ratio', 'created_at', 'description', 'draft_status', 'inspired_by_url',
       'thumbnail_url', 'thumbnail_alt_text', 'total_time_minutes', 'video_url', 'updated_at', 'credits',
       'approved_at', 'is_one_top', 'servings_noun_plural', 'renditions', 'beauty_url', 'is_shoppable', 'topics',
       'video_ad_content', 'seo_title', 'yields', 'original_video_url', 'canonical_id', 'cook_time_minutes',
       'promotion', 'video_id', 'user_ratings.count_positive', 'user_ratings.score', 'user_ratings.count_negative',
       'price.total', 'price.updated_at', 'price.portion', 'price.consumption_total', 'price.consumption_porti

You will need to pass these `headers` with every HTTP request to the API. The API key (X-RapidAPI-Key) is how the server keeps track of how many requests you have made.

Take a look at [the documentation](https://rapidapi.com/apidojo/api/tasty). We will use the recipes/list endpoint.

Make sure you are logged into the account you are created, and select the recipes/list endpoint from the menu at left. Notice that this brings up a form that you can fill in, which generates the corresponding code.  By default, it provides Node.js code; change this to Python with the Requests client.

1\. Search for recipes containing "daikon" (an Asian radish) and request the JSON data, and convert it to a Pandas data frame.

In [38]:
import pandas as pd

# Convert the API response to JSON
data_daikon = response.json()

df_daikon = pd.json_normalize(
    data_daikon["results"]
)

df_daikon

,nutrition_visibility,country,instructions,keywords,facebook_posts,language,seo_path,id,brand,slug,...,nutrition.fiber,nutrition.updated_at,nutrition.protein,nutrition.fat,nutrition.calories,nutrition.sugar,show.name,show.id,total_time_tier.tier,total_time_tier.display_tier
0,auto,US,"[{'start_time': 9666, 'appliance': None, 'end_...",None,[],eng,"8757513,9295873,64461",4490,None,top-chef-junior-pork-banh-mi-burger,...,8,2024-09-03T03:10:44+02:00,41,48,905,34,Tasty,17,under_30_minutes,Under 30 minutes
1,auto,US,"[{'start_time': 423000, 'appliance': None, 'ha...",,[],eng,"8757513,9295873,64454",7224,None,japanese-omelette,...,10,2024-05-01T08:03:22+02:00,28,22,412,8,Tasty 101,63,under_15_minutes,Under 15 minutes
2,auto,US,"[{'start_time': 10833, 'appliance': None, 'hac...","classic vietnamese dish, hanger steak, homemad...",[],eng,"8757513,9295873,64461",5375,None,bun-bo-xa-ot-vietnamese-steak-and-rice-noodle-...,...,45,2024-05-01T08:05:56+02:00,43,39,1316,65,Tasty,17,under_30_minutes,Under 30 minutes
3,auto,US,"[{'start_time': 0, 'appliance': None, 'end_tim...","easy weeknight salmon, ginger rice, healthy sa...",[],eng,None,9617,None,instant-pot-miso-butter-salmon-and-ginger-rice,...,6,2026-01-11T07:00:49+01:00,42,27,612,7,Tasty,17,under_30_minutes,Under 30 minutes


In [39]:
df_daikon.shape

(4, 68)

2\. How many recipes containing daikon are there? Which one is the cheapest per portion?

There are four recipes containing daikon. The cheapest per portion is $650

In [ ]:
len(df_daikon)

df_daikon[
    ["name", "price.portion"]].sort_values("price.portion").head(1)

,name,price.portion
0,Top Chef Junior Pork Bánh Mì Burger,650.0


In [ ]:
print("Number of daikon recipes:", len(df_daikon))

df_daikon[
    ["name", "price.portion"]].sort_values("price.portion").head(1)

Number of daikon recipes: 4


,name,price.portion
0,Top Chef Junior Pork Bánh Mì Burger,650.0


3\. Find recipes containing avocado and request the JSON data.

**Hint:** Note that there are hundreds of results, but the API only returns 20 results by default and only 40 results maximum, even if you specify the `size=` parameter, so you will need to use a `for` loop, incrementing the `from=` parameter. Be sure to respect the API's rate limits (or you may be blocked!)

**Note:** Recall from the example in the reading that we created an empty list `episodes = []` and then added the results of each request to it in the for loop with `episodes.extend(response.json())`. But note that in the recipes/list endpoint there are two keys: count and results. We only want the results so try instead `.extend(response.json()).get("results")`.

**Suggestion:** Try writing a loop that only makes 2 or 3 requests first so you can test that it's working correctly. Also, make sure you use separate cells for code that makes requests to the API versus processing of the results; you don't want to rerun the requests unless absolutely necessary!

In [42]:
import time

avocado_recipes = []

for start in range(0, 120, 40):

    querystring = {
        "q": "avocado",
        "size": "40",
        "from": str(start)
    }

    response = requests.get(
        url,
        headers=headers,
        params=querystring
    )

    avocado_recipes.extend(
        response.json().get("results")
    )

    time.sleep(0.5)

len(avocado_recipes)

120

In [44]:
querystring = {
    "q": "avocado",
    "size": "40",
    "from": "0"
}

response = requests.get(
    url,
    headers=headers,
    params=querystring
)

response.json()["count"]

444

In [45]:
# run full loop for all 444 recipes
total_avocado = response.json()["count"]

avocado_recipes = []

for start in range(0, total_avocado, 40):

    querystring = {
        "q": "avocado",
        "size": "40",
        "from": str(start)
    }

    response = requests.get(
        url,
        headers=headers,
        params=querystring
    )

    avocado_recipes.extend(
        response.json().get("results")
    )

    time.sleep(0.5)

len(avocado_recipes)

444

In [47]:
df_avocado = pd.json_normalize(avocado_recipes)

df_avocado.shape

(444, 77)

4\. For the recipes containing avocado, compute the proportion of reviews that are positive. For the avocado recipes with over 500 reviews, which one has the highest proportion of positive reviews?


The recipe with the highest proportion of positive reviews is the grilled salmon

In [48]:
df_avocado["total_reviews"] = (
    df_avocado["user_ratings.count_positive"]
    + df_avocado["user_ratings.count_negative"])

df_avocado["positive_proportion"] = (
    df_avocado["user_ratings.count_positive"]
    / df_avocado["total_reviews"])

df_avocado[
    df_avocado["total_reviews"] > 500][
    ["name", "total_reviews", "positive_proportion"]].sort_values(
    "positive_proportion",
    ascending=False).head(1)

,name,total_reviews,positive_proportion
13,Grilled Salmon With Avocado Salsa,1898.0,0.987355


5\. Take the avocado JSON data from above (you do NOT need to read in the data from the REST API again). How many recipes are vegetarian? You should be able to identify this from the "tags" attribute.

Hint: Try using `json_normalize` with "tags" as the record path to flatten the data so that there is one row for each tag.

In [49]:
df_tags = pd.json_normalize(
    avocado_recipes,
    record_path="tags",
    meta=["name"],
    meta_prefix="recipe_")

df_tags

,root_tag_type,name,id,parent_tag_name,display_name,type,recipe_name
0,cuisine,north_american,64444,cuisine,North American,cuisine,Avocado Carbonara
1,cuisine,italian,64453,european,Italian,european,Avocado Carbonara
2,cooking_style,comfort_food,64462,cooking_style,Comfort Food,cooking_style,Avocado Carbonara
3,healthy,healthy,64466,None,Healthy,healthy,Avocado Carbonara
4,difficulty,easy,64471,difficulty,Easy,difficulty,Avocado Carbonara
...,...,...,...,...,...,...,...
11757,meal,beef,9299515,dinner,Beef,dinner,Cemitas Poblanas
11758,ingredients_to_avoid,eggs,10623588,ingredients_to_avoid,Eggs,ingredients_to_avoid,Cemitas Poblanas
11759,ingredients_to_avoid,gluten,10623589,ingredients_to_avoid,Gluten,ingredients_to_avoid,Cemitas Poblanas
11760,ingredients_to_avoid,dairy,10623608,ingredients_to_avoid,Dairy,ingredients_to_avoid,Cemitas Poblanas


In [50]:
df_tags[
    df_tags["name"] == "vegetarian"]["recipe_name"].nunique()

206

206 avocado recipes are vegetarian